### PR TITLE
Speed up on_trait_change, part IV

### PR DESCRIPTION
**Pull request by mdickinson**
This PR completes the removal of `HasTraits` machinery from the traits minilanguage parsing machinery.

### CODE REVIEW

**Comment by mdickinson** [comment]
Some timings using the following script:
<details>

```python
import time

from traitsui.api import Group
from traitsui.group import ShadowGroup


trials = 1000

# Warm up
for _ in range(trials):
    ShadowGroup(shadow=Group())

start = time.perf_counter()
for _ in range(trials):
    ShadowGroup(shadow=Group())
time_taken = time.perf_counter() - start
print(f"Time to create {trials} ShadowGroups: {time_taken:.3f}")
```

On this branch:
```
Time to create 1000 ShadowGroups: 1.203
```

On main:
```
Time to create 1000 ShadowGroups: 1.392
```

On main, prior to #1492, #1491 and #1490 (commit 52f4d00248cec5dbf0826de4e846b4ad83cf072e):
```
Time to create 1000 ShadowGroups: 2.199
```

File: traits/traits_listener.py [hunk 1]
```diff
@@ -170,57 +184,81 @@ def handle_dict_items(self, object, name, old, new):
...
-    #: (*name*, *type*) listener pairs, where *type* defines the type of
-    #: listener, one of: (SIMPLE_LISTENER, LIST_LISTENER, DICT_LISTENER).
-    active = Instance(WeakIDKeyDict, ())
+    def __init__(
+        self,
+        *,
+        name,
+        metadata_name="",
+        metadata_defined=True,
+        handler=None,
```
Comment (mdickinson) [inline traits/traits_listener.py:236]: `handler` and `wrapped_handler_ref` should be required, not optional, but there are currently some unit tests that don't pass them.  We can make these required in another PR that updates the test code accordingly.

File: traits/traits_listener.py [hunk 2]
```diff
@@ -1122,10 +1194,11 @@ def parse_item(self, *, terminator, deferred, handler_type):
                 last = result
                 while last.next is not None:
                     last = last.next
-                last.next = lg = ListenerGroup(items=[next, result])
+                lg = ListenerGroup(items=[next, result])
```
Comment (mdickinson) [inline traits/traits_listener.py:1197]: General comment: there are a number of simplifications that would be possible if we didn't have to deal with  recursion and cycles; the cycles add substantial complication for a feature that's rarely used. Leaving the cycles out of the new `observe` machinery seems as though it was a Good Thing.

**Review by corranwebster — APPROVED** [review]
I am comfortable with this change - the only use of traits was to support the `Property` traits, it would seem.

Do you want to make `ListenerBase` a proper `abc.ABC` with `@abstractmethods`?

**Comment by mdickinson** [comment]
> Do you want to make `ListenerBase` a proper `abc.ABC` with `@abstractmethods`?

It's tempting, but once I start down that route there are so many refactorings and simplifications I'd _like_ to make to this code that it's hard to stop (I rolled several such changes back before making this PR). :-) I think I'd rather not make further changes to what should be being considered legacy code now that we've got the performance benefit we were looking for.

File: traits/traits_listener.py [hunk 3]
```diff
@@ -124,14 +121,31 @@ def not_event(value):
...
+        ----------
+        notify : bool
+            True if this listener should notify, else False.
+        """
+        raise NotImplementedError
 
-    # Should changes to this item generate a notification to the handler?
-    # notify = Bool
+    def set_next(self, next):
+        """ Set the child listener for this one.
```
Comment (rahulporuri) [inline traits/traits_listener.py:140]:
```suggestion
        """ Set the child listener for this listener.
```

**Review by rahulporuri — APPROVED** [review]
